### PR TITLE
[VSTS] Create a script that will allow to merge the xml results.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

File: tools/unit-tests-merge.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+using System.Text;
+using System.Xml;
+
+var args = Environment.GetCommandLineArgs ();
+// first arg is "/Library/Frameworks/Mono.framework/Versions/4.8.0/lib/mono/4.5/csharp.exe"
+// second arg the script itself
+// then comes the ones we care about, the path with all the xunit and nunit files
+
+if (args.Length < 3) {
+    Console.WriteLine ($"Missing path for the direcotry that contains the test results files.");
```
Comment (rolfbjarne) [inline tools/unit-tests-merge.cs]:
```suggestion
    Console.WriteLine ($"Missing path for the directory that contains the test results files.");
```

File: tools/unit-tests-merge.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    // create a writer for the final solution
+    var outPutPath = "tests-grouped-result.xml";
+
+    if (File.Exists (outPutPath))
+        File.Delete (outPutPath);
+
+    using (var writer = File.CreateText (outPutPath)) {
+        writer.WriteLine ("<?xml version=\"1.0\" encoding=\"utf-8\"?>");
+        writer.WriteLine ($"<test-results name=\"Test results\" errors=\"{errors}\" inconclusive=\"{inconclusive}\" ignored=\"{ignored}\" invalid=\"{invalid}\" not-run=\"{notRun}\" date=\"{date}\" time=\"{time.Value}\" total=\"{total}\" failures=\"{failures}\" skipped=\"{skipped}\">");
+        writer.WriteLine ("<environment os-version=\"unknown\" platform=\"unknown\" cwd=\"unknown\" machine-name=\"unknown\" user=\"unknown\" user-domain=\"unknown\" nunit-version=\"xUnit.net 2.4.0.4049\" clr-version=\"64-bit .NET Standard [collection-per-class, non-parallel]\"></environment>");
```
Comment (rolfbjarne) [inline tools/unit-tests-merge.cs]: Can you get all the `unknown`s from any of the input xmls?

Comment (mandel-macaque) [reply tools/unit-tests-merge.cs]: @rolfbjarne that is something I did not like, and unfortunately, I cannot. But I'm going to fix the runner to report that. I want to see the environment and include the touch args used. We will need to do another PR (we first need to see if VSTS is happy with the HUGE file is going to generate.

File: tools/unit-tests-merge.cs (hunk 1 above)
Comment (mandel-macaque) [reply tools/unit-tests-merge.cs]: Me and my dyslexia (had to autocorrect it here :P )

File: tools/unit-tests-merge.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env /Library/Frameworks/Mono.framework/Commands/csharp
```
Comment (spouliot) [inline tools/unit-tests-merge.cs]: we normally use `.csharp` for script files, so they are not confused with, to be compiled, `.cs` files

File: tools/unit-tests-merge.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+using System.Text;
+using System.Xml;
+
+var args = Environment.GetCommandLineArgs ();
+// first arg is "/Library/Frameworks/Mono.framework/Versions/4.8.0/lib/mono/4.5/csharp.exe"
+// second arg the script itself
+// then comes the ones we care about, the path with all the xunit and nunit files
+
+if (args.Length < 3) {
+    Console.WriteLine ($"Missing path for the directory that contains the test results files.");
```
Comment (spouliot) [inline tools/unit-tests-merge.cs]: formatting seems incorrect (spaces)

File: tools/unit-tests-merge.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    var errors = 0L;
+    var inconclusive = 0L;
+    var ignored = 0L;
+    var invalid = 0L;
+    var notRun = 0L;
+    var total = 0L;
+    var failures = 0L;
+    var skipped = 0L;
+    var date = DateTime.Today;
+    TimeSpan? time = null;
```
Comment (spouliot) [inline tools/unit-tests-merge.cs]: can't you set it to `Zero` and just add to it later ? (skipping `?` and the `HasValue` check in a loop?)

File: tools/unit-tests-merge.cs [hunk 6]
```diff
@@ -103,27 +106,43 @@
...
+		writer.WriteLine ("    </results>");
+		writer.WriteLine ("  </test-suite>");
 		writer.WriteLine ("</test-results>");
 	}
 	Console.WriteLine ($"File merge completed to file {outPutPath}");
 
 } catch (Exception e) {
 	Console.WriteLine ($"Got {e.Message}");
+	throw e;
 	Environment.Exit (-1);
```
Comment (spouliot) [inline tools/unit-tests-merge.cs]: with the `throw` you'll never return `-1`

File: tools/unit-tests-merge.cs [hunk 7]
```diff
@@ -103,27 +106,43 @@
...
-		writer.WriteLine ($"<environment os-version=\"{os_version}\" platform=\"{platform}\" cwd=\"{cwd}\" machine-name=\"{machine_name}\" user=\"{user}\" user-domain=\"{user_domain}\" nunit-version=\"{nunit_version}\" clr-version=\"{clr_version}\"></environment>");
-		writer.WriteLine ("<culture-info current-culture=\"unknown\" current-uiculture=\"unknown\" />");
+		writer.WriteLine ($"  <environment os-version=\"{os_version}\" platform=\"{platform}\" cwd=\"{cwd}\" machine-name=\"{machine_name}\" user=\"{user}\" user-domain=\"{user_domain}\" nunit-version=\"{nunit_version}\" clr-version=\"{clr_version}\"></environment>");
+		writer.WriteLine ("  <culture-info current-culture=\"unknown\" current-uiculture=\"unknown\" />");
+		writer.WriteLine ($"  <test-suite type=\"Assemblies\" name=\"Device Tests\" executed=\"True\" success=\"{(errors == 0 && failures == 0)? "True" : "False" }\" result=\"{(errors == 0 && failures == 0)? "Success" : "Failure"}\" time=\"{time.Value}\">");
+    	writer.WriteLine ("    <results>");
```
Comment (spouliot) [inline tools/unit-tests-merge.cs]: minor: indent

**Pull request closed without merging**